File: web-app/src/test/utils/profanityFilter.test.ts
import { describe, it, expect } from 'vitest';
import { containsProfanity, sanitise } from '../../utils/profanityFilter';

describe('containsProfanity', () => {
  it('returns false for clean text', () => {
    expect(containsProfanity('Hello, how are you?')).toBe(false);
  });

  it('returns false for empty string', () => {
    expect(containsProfanity('')).toBe(false);
  });

  it('detects a known bad word', () => {
    expect(containsProfanity('you are an ass')).toBe(true);
  });

  it('is case-insensitive', () => {
    expect(containsProfanity('ASS')).toBe(true);
  });

  it('does not flag partial matches inside clean words', () => {
    // "class" contains "ass" but with word boundaries it should be fine
    expect(containsProfanity('I am in class')).toBe(false);
  });
});

describe('sanitise', () => {
  it('replaces a bad word with asterisks', () => {
    const result = sanitise('you are an ass');
    expect(result).not.toContain('ass');
    expect(result).toContain('***');
  });

  it('returns clean text unchanged', () => {
    expect(sanitise('Hello there')).toBe('Hello there');
  });
});

File: android-app/app/src/test/java/com/djtaylor/cartalk/util/ProfanityFilterTest.kt
package com.djtaylor.cartalk.util

import org.junit.Assert.*
import org.junit.Test

class ProfanityFilterTest {

    @Test
    fun `clean text is not flagged`() {
        assertFalse(ProfanityFilter.containsProfanity("Your brake light is out"))
        assertFalse(ProfanityFilter.containsProfanity("Hey I accidentally bumped your car"))
        assertFalse(ProfanityFilter.containsProfanity(""))
    }

    @Test
    fun `banned word in text is detected`() {
        assertTrue(ProfanityFilter.containsProfanity("what the fuck"))
        assertTrue(ProfanityFilter.containsProfanity("You're an asshole"))
    }

    @Test
    fun `detection is case insensitive`() {
        assertTrue(ProfanityFilter.containsProfanity("What THE FUCK"))
        assertTrue(ProfanityFilter.containsProfanity("SHIT happens"))
    }

    @Test
    fun `word boundary prevents false positive`() {
        // "scunthorpe problem" — should NOT match 'cunt' inside 'Scunthorpe'
        // Our regex uses word boundaries so this is a known trade-off to document
        // For CarTalk we accept possible false positives on partial matches as a safety measure
        // Test that an isolated banned word IS caught
        assertTrue(ProfanityFilter.containsProfanity("you are a cunt"))
    }

    @Test
    fun `sanitise replaces banned words with asterisks`() {
        val result = ProfanityFilter.sanitise("what the fuck is this")
        assertFalse(result.contains("fuck", ignoreCase = true))
        assertTrue(result.contains("****"))
    }

    @Test
    fun `sanitise preserves clean words`() {
        val input = "Your brake light is out"
        assertEquals(input, ProfanityFilter.sanitise(input))
    }

    @Test
    fun `contact info sharing is allowed`() {
        assertFalse(ProfanityFilter.containsProfanity("My number is 123-456-7890 email blah@example.com"))
    }
}
